### DISCUSSION TITLE
Account Abstraction - entryPoint address

### DISCUSSION THREAD

**Discussion by SiewLin-AIDeaBuddy** [Cyfrin/foundry-full-course-cu#1950] [Q&A] [answered]
While going through the account-abstraction section, on the [HelperConfig](https://github.com/Cyfrin/minimal-account-abstraction/blob/321cc2a2d1d4f58a38feca6580a95584aeb2c060/script/HelperConfig.s.sol#L67-L107), there were these codes :

```javascript

    function getEthMainnetConfig() public pure returns (NetworkConfig memory) {
        // This is v7
        return NetworkConfig({
            entryPoint: 0x0000000071727De22E5E9d8BAf0edAc6f37da032,
            usdc: 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48,
            account: BURNER_WALLET
        });
        // https://blockscan.com/address/0x0000000071727De22E5E9d8BAf0edAc6f37da032
    }

    function getEthSepoliaConfig() public pure returns (NetworkConfig memory) {
        return NetworkConfig({
            entryPoint: 0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789,
            usdc: 0x53844F9577C2334e541Aec7Df7174ECe5dF1fCf0, // Update with your own mock token
            account: BURNER_WALLET
        });
    }

    function getArbMainnetConfig() public pure returns (NetworkConfig memory) {
        return NetworkConfig({
            entryPoint: 0x0000000071727De22E5E9d8BAf0edAc6f37da032,
            usdc: 0xaf88d065e77c8cC2239327C5EDb3A432268e5831,
            account: BURNER_WALLET
        });
    }
…
```
I have a few questions regarding the entryPoint addresses :
1. Where do we get references from for the entryPoint addresses for the various network configs? 
2. How do we know which address to use and if the addresses are indeed the official or latest version of implementation?
3. How to know if the same entry point address can be used on various chains or available only on specific chains ( is it purely based on the `Multichain Portfolio` info tab in etherscan?)
4. And when do we need/know if to update the entryPoint to latest official version of deployed address especially for mainnet deployment test run? ( based on `proxy upgradable` info in etherscan? If not proxy upgradable, how do we get timely release of new entryPoint address? )

As these info are hard-coded addresses, would appreciate if someone can help to explain in more details to clear the doubts above. Thanks.

**Comment by EngrPips** [comment]
> While going through the account-abstraction section, on the [HelperConfig](https://github.com/Cyfrin/minimal-account-abstraction/blob/321cc2a2d1d4f58a38feca6580a95584aeb2c060/script/HelperConfig.s.sol#L67-L107), there were these codes :
> 
> ```javascript
> 
>     function getEthMainnetConfig() public pure returns (NetworkConfig memory) {
>         // This is v7
>         return NetworkConfig({
>             entryPoint: 0x0000000071727De22E5E9d8BAf0edAc6f37da032,
>             usdc: 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48,
>             account: BURNER_WALLET
>         });
>         // https://blockscan.com/address/0x0000000071727De22E5E9d8BAf0edAc6f37da032
>     }
> 
>     function getEthSepoliaConfig() public pure returns (NetworkConfig memory) {
>         return NetworkConfig({
>             entryPoint: 0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789,
>             usdc: 0x53844F9577C2334e541Aec7Df7174ECe5dF1fCf0, // Update with your own mock token
>             account: BURNER_WALLET
>         });
>     }
> 
>     function getArbMainnetConfig() public pure returns (NetworkConfig memory) {
>         return NetworkConfig({
>             entryPoint: 0x0000000071727De22E5E9d8BAf0edAc6f37da032,
>             usdc: 0xaf88d065e77c8cC2239327C5EDb3A432268e5831,
>             account: BURNER_WALLET
>         });
>     }
> …
> ```
> I have a few questions regarding the entryPoint and usdc addresses :
> 1. Where do we get references from for the entryPoint and the usdc addresses for the various network configs? 
> 2. How do we know which address to use and if the addresses are indeed the official or latest version of implementation?
> 3. How to know if the same entry point address can be used on various chains or available only on specific chains ( is it purely based on the `Multichain Portfolio` info tab in etherscan?)
> 4. And when do we need/know if to update the entryPoint to latest official version of deployed address especially for mainnet deployment test run? ( based on `proxy upgradable` info in etherscan? If not proxy upgradable, how do we get timely release of new entryPoint address? )
> 
> As these info are hard-coded addresses, would appreciate if someone can help to explain in more details to clear the doubts above. Thanks.

I am yet to take this section of the course as of now so I might not be very helpful but If you haven't finished the section  I will say you should keep going as Patrick might later touch on all of this questions before the end of the course. Anyways we can look forward to  @PatrickAlphaC help as soon as they are available.

**Comment by SiewLin-AIDeaBuddy** [comment]
I find this QA link helps to clear part of the doubts :
https://ethereum.stackexchange.com/questions/153472/how-can-i-identify-the-blockchain-where-a-smart-contract-is-deployed-using-its-a

but still have no idea where/how I can get the latest/verfied deployment address of entryPoint if it is not provided in the additional comment `// https://blockscan.com/address/0x0000000071727De22E5E9d8BAf0edAc6f37da032` provided in the `HelperConfig` file. I also couldn't find it through the [ERC-4337 doc](https://eips.ethereum.org/EIPS/eip-4337#entrypoint-definition). In etherscan, we can't seem to query based on the name tag like `ERC-4337 Entry Point`...

**Reply by EngrPips** [reply]
Thanks for sharing this with us. I am looking to take the `Account Abstraction` course soon enough, and then I will come back here to share whatever I find.

**Comment by save-the-planet08** [comment]
Hi, I actually asked myself the same question!

That's why I simply searched for this exact address on etherscan. Here I was able to see exactly what this contract does for "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789" and exactly what it has to do with account abstraction (EIP-4337). 

However, I haven't checked exactly whether this contract is now upgradable or remains on the chain forever. 

I hope I was able to help you.

Here are the links:
https://etherscan.io/address/0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789#code,
https://vscode.blockscan.com/ethereum/0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789

**Reply by EngrPips** [reply]
Interesting. Good heads-up. can't wait to take the course.

**Reply by SiewLin-AIDeaBuddy** [reply]
My question is more on how we even know what entryPoint address to query in etherscan/arbitrum (or any other supporting chains) in the first place, where the official/reliable source we should refer if it is not hard-coded/given in the HelperConfig file for us.

For example when there's a new version release of EntryPoint Contract v0.8, v0.9, v0.10 ...etc, where can we refer the source about the release and its associated official deployment contract address.. I couldn't seem to find them in the [doc](https://eips.ethereum.org/EIPS/eip-4337#entrypoint-definition) and the [github](https://github.com/ethereum/ercs/blob/master/ERCS/erc-4337.md). Therefore, curious where else can we get such important reference from.

**Reply by EngrPips** [reply]
Now it is becoming more clear to me how genuine your curiosity it.

**Comment by PatrickAlphaC** [comment, marked as answer]
> Where do we get references from for the entryPoint addresses for the various network configs?

Technically, you can use any EntryPoint address you want! We as a community have decided to use the ones created by [this group](https://github.com/eth-infinitism/account-abstraction/) as the Ethereum foundation put out a grant for someone to deploy the codebase, and they were the ones who did it. I had to check their deployments in order to confirm, I asked them to update their docs with proper addresses, but they have not responded :/ 

> How do we know which address to use and if the addresses are indeed the official or latest version of implementation?

See the above answer ^. The `EntryPoint.sol` can be any contract that the alt-mempool nodes decide to use! 

> How to know if the same entry point address can be used on various chains or available only on specific chains ( is it purely based on the Multichain Portfolio info tab in etherscan?)

If someone used `CREATE2` to deploy the contract, it _could_ be the same address on different chains, but it's unsafe to think there is a 100% chance that it's the same address. Remember, it's a deployed smart contract so it could be assigned different addresses! 

> And when do we need/know if to update the entryPoint to latest official version of deployed address especially for mainnet deployment test run? ( based on proxy upgradable info in etherscan? If not proxy upgradable, how do we get timely release of new entryPoint address? )

It sort of doesn't matter! Your account can/will work with any version of entrypoint so long as it conforms to the standards. Only the alt-mempool nodes need to care about what `EntryPoint.sol` to use, since they are the ones sending the `handleOps` function call. 

> As these info are hard-coded addresses, would appreciate if someone can help to explain in more details to clear the doubts above. 

Yep! Great question!

It's sort of the same as "What is the official Uniswap address?", well, you're probably looking for the one deployed by "Uniswap Labs", but technically, you could use any contract deployed by anyone that has the Uniswap code.

**Reply by SiewLin-AIDeaBuddy** [reply]
As outlined in Patrick's answer, for any deployed smart contracts, they could be assigned different addresses. Therefore it could be confusing or doubtful if there is no published info on their repo/project page to confirm the deployment that leads the user more directly to know if the contract or its owner is the intended rightful target of interaction, though it's the alt-mempool to decide which entry point to implement, as well as to understand/have an overview of the entire workflow what interactions happened in-between these processes [ something like same codes same logic but different addresses, then why use B instead of A... any state / data context behind the scene of application...etc ]

Uniswap Labs does outline the deployment contract addresses for different chains [ref](https://docs.uniswap.org/contracts/v3/reference/deployments/) so it's more convincing to take it as legitimate/official addresses. Hopefully the same deployment status also being published by the mentioned group for ease of reference and its usability.

Thanks @PatrickAlphaC  for taking the efforts to attend these questions.
